src/types/mutex.rs

```diff
@@ -0,0 +1,375 @@
+use std::cell::UnsafeCell;
+use std::ops::{Deref, DerefMut};
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    LockResult, PoisonError,
+};
+use std::thread;
+
+// See std::sync::poison in the rust standard library.
+// This is more-or-less copied from there since it is not public.
+// this type detects a panic and poisons the wrapping mutex
+pub(crate) struct Flag {
```

Comment:
Think this can be private to this file?

```suggestion
struct Flag {
```